File: code/Graphs/dijkstra.py
from heapq import *
def dijkstra(n, root, g):  # g = {node: (cost, neigh)}
	dist = [float("inf")]*n
	dist[root] = 0
	prev = [-1]*n

	pq = [(0, root)]
	heapify(pq)
	visited = set([])

	while len(pq) != 0:
		_, node = heappop(pq)

		if node in visited: continue
		visited.add(node)

		# In case of disconnected graphs
		if node not in g:
			continue

		for cost, neigh in g[node]:
			alt = dist[node] + cost
			if alt < dist[neigh]:
				dist[neigh] = alt
				prev[neigh] = node
				heappush(pq, (alt, neigh))
	return dist
	

line = input()
while line != "0 0 0 0":
	n, m, q, s = map(int, line.split())

	g = {}
	for i in range(n):
		g[i] = [(0, i)]

	for i in range(m):
		u, v, w = map(int, input().split())
		g[u].append((w, v))

	
	dist = dijkstra(n, s, g)
	for _ in range(q):
		task = int(input())
		print(dist[task] if dist[task] < float("inf") else "Impossible")

	line = input()

File: code/Graphs/dijkstra.cpp
#include "header.h"
vector<int> dijkstra(int n, int root, map<int, vector<pair<int, int>>>& g) {
	unordered_set<int> visited;
	vector<int> dist(n, INF);
    priority_queue<pair<int, int>> pq;
    dist[root] = 0;
    pq.push({0, root});
    while (!pq.empty()) {
        int node = pq.top().second;
        int d = -pq.top().first;
        pq.pop();

        if (in(node, visited)) continue;
        visited.insert(node);

        // In case of disconnected graphs
        if (not in(node, g)) continue;

        for (auto e : g[node]) {
            int neigh = e.first;
            int cost = e.second;
            if (dist[neigh] > dist[node] + cost) {
                dist[neigh] = dist[node] + cost;
                pq.push({-dist[neigh], neigh});
            }
        }
    }
    return dist;
}


int main() {

    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m, q, s;
    cin >> n >> m >> q >> s;
    while (!(n == 0 && m == 0 && q == 0 && s == 0)) {
        map<int, vector<pair<int, int>>> g;
        for (int i = 0; i < m; i++) {
            int u, v, w;
            cin >> u >> v >> w;
            g[u].push_back({v, w});
        }

        set<int> queries;
        vector<int> queries_help;
        int q_help = 0;
        while (q--) {
            int v;
            cin >> v;
            queries.insert(v);
            queries_help.push_back(v);
            q_help++;

        }

        
		vector<int> dist = dijkstra(n, s, g);
        for (auto query: queries_help) {
            if (dist[query] == INF) cout << "Impossible" << endl;
            else cout << dist[query] << endl;
        }

        cin >> n >> m >> q >> s;
    }

    return 0;
}